src/rust-util/pelikan/src/protocol/mod.rs

```diff
@@ -14,9 +14,49 @@
 
 #[cfg(feature = "protocol_admin")]
 pub mod admin;
+#[cfg(feature = "protocol_ping")]
+pub mod ping;
+
+use std::error::Error;
+
+use ccommon::buf::OwnedBuf;
+
+/// An error that could indicate that there weren't enough
+/// bytes to successfully parse the buffer.
+pub trait PartialParseError {
+    /// Indicates that this error occurred because there
+    /// weren't enough bytes in the buffer.
+    fn is_unfinished(&self) -> bool;
+}
+
+pub trait QuitRequest {
+    fn is_quit(&self) -> bool;
+}
+
+/// A type that can be serialized/deserialized.
+///
+/// TODO: Name this better
+pub trait Serializable: Sized {
+    type ParseError: Error + PartialParseError;
+    type ComposeError: Error;
+
+    fn reset(&mut self);
+
+    fn parse(&mut self, buf: &mut OwnedBuf) -> Result<(), Self::ParseError>;
```

Comment:
My only issue with this here is that it'll add another 6 required (since we'll need to split into `Serialize`, `Deserialize` and something that has `reset`) trait impls for each protocol. We're already at 5 different ones.

This isn't a big objection - if you think having the extra flexibility is worth it then I'll split the traits.

Reply:
Let's revisit and discuss later.